source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -742,29 +745,43 @@ static ProcessingMode allDisabledMode() {
 }
 
 void Filter::mergePerRouteConfig() {
-  auto&& merged_config = Http::Utility::getMergedPerFilterConfig<FilterConfigPerRoute>(
+  if (route_config_merged_) {
+    return;
+  }
+  route_config_merged_ = true;
+
+  auto merged_config = Http::Utility::getMergedPerFilterConfig<FilterConfigPerRoute>(
       decoder_callbacks_,
       [](FilterConfigPerRoute& dst, const FilterConfigPerRoute& src) { dst.merge(src); });
-  if (!merged_config) {
-    return;
+  if (merged_config.has_value()) {
+    if (merged_config->disabled()) {
+      // Rather than introduce yet another flag, use the processing mode
+      // structure to disable all the callbacks.
+      ENVOY_LOG(trace, "Disabling filter due to per-route configuration");
+      const auto all_disabled = allDisabledMode();
+      decoding_state_.setProcessingMode(all_disabled);
+      encoding_state_.setProcessingMode(all_disabled);
+      return;
+    }
+    if (merged_config->processingMode()) {
+      ENVOY_LOG(trace, "Setting new processing mode from per-route configuration");
+      decoding_state_.setProcessingMode(*(merged_config->processingMode()));
+      encoding_state_.setProcessingMode(*(merged_config->processingMode()));
+    }
+    if (merged_config->grpcService()) {
+      ENVOY_LOG(trace, "Setting new GrpcService from per-route configuration");
+      grpc_service_ = *merged_config->grpcService();
+    }
   }
-  if (merged_config->disabled()) {
-    // Rather than introduce yet another flag, use the processing mode
-    // structure to disable all the callbacks.
-    ENVOY_LOG(trace, "Disabling filter due to per-route configuration");
+
+  // Disable the filter if there is no valid GrpcService configuration to avoid possible runtime
+  // panics.
+  if (grpc_service_.target_specifier_case() ==
+      envoy::config::core::v3::GrpcService::TargetSpecifierCase::TARGET_SPECIFIER_NOT_SET) {
+    ENVOY_LOG(trace, "Disabling filter due to no valid GrpcService configuration");
     const auto all_disabled = allDisabledMode();
```

Comment:
Sound reasonable to me. I will update it.

Reply:
Agreed that this is most likely a config error. But I think the existing `failure_mode_allow` applies to this case, specifically for `if the gRPC stream cannot be established`:

```
  // By default, if the gRPC stream cannot be established, or if it is closed
  // prematurely with an error, the filter will fail. Specifically, if the
  // response headers have not yet been delivered, then it will return a 500
  // error downstream. If they have been delivered, then instead the HTTP stream to the
  // downstream client will be reset.
  // With this parameter set to true, however, then if the gRPC stream is prematurely closed
  // or could not be opened, processing continues without error.
  bool failure_mode_allow = 2;
```

Reply:
@ggreenway In the latest commit, if there is a invalid grpc service, I will treat it as grpc error directly. And if the `failure_mode_allow` is set to true, the filter chain will continue the processing, if the `faliure_mode_allow` is set to false, the `sendLocalReply` will be called.

Reply:
This does seem like it would be a configuration error, should we just fail config validation if this isn't set? The filter would be useless without this set right?

Reply:
I do wonder if it would make things a bit nicer if we made grpc_service required on the filter config to avoid this scenario? I think that would ensure that there always is a grpc_service present, and users can still change whether its enabled by modifying processing mode

Reply:
> I do wonder if it would make things a bit nicer if we made grpc_service required on the filter config to avoid this scenario? I think that would ensure that there always is a grpc_service present, and users can still change whether its enabled by modifying processing mode

Yeap. It would be a best way. But it would be a break change that some exist users may be impacted.

Reply:
I checked the metadata of `ext_proc`. It still be marked as `alpha` status. May be it's possible to add this new validation to its proto API?

Reply:
cc @snowp

Reply:
Yeah I think that makes sense to me, unless there is any objection from @gbrail or from anyone using this at Google. Not sure who would be the best to answer for Google, @KBaichoo do you know?

Reply:
+1 I agree the filter config check to made grpc_service required is cleaner and nicer. I think @mpwarres team from Google is using ext_proc. What is your opinion on this?

Reply:
cc @stevenzzzz @yanjunxiang-google  who are also in this area at google